teuthology/provision/openstack.py

```diff
@@ -57,52 +57,78 @@ def init_user_data(self, os_type, os_version):
+                    break
+                except subprocess.CalledProcessError:
+                    log.warning("openstack add volume failed unexpectedly; retrying")
+
     def attach_volumes(self, name, volumes):
         """
         Create and attach volumes to the named OpenStack instance.
         """
         for i in range(volumes['count']):
             volume_name = name + '-' + str(i)
-            try:
-                OpenStack().run("volume show -f json " +
-                                volume_name)
-            except subprocess.CalledProcessError as e:
-                if 'No volume with a name or ID' not in e.output:
-                    raise e
-                # do not use OpenStack().run because its
-                # bugous for volume create as of openstackclient 3.2.0
-                # https://bugs.launchpad.net/python-openstackclient/+bug/1619726
-                misc.sh(
-                    "openstack volume create -f json " +
-                    config['openstack'].get('volume-create', '') + " " +
-                    " --property ownedby=" + config.openstack['ip'] +
-                    " --size " + str(volumes['size']) + " " +
-                    volume_name)
-            with safe_while(sleep=2, tries=100,
+            with safe_while(sleep=10, tries=2,
                             action="volume " + volume_name) as proceed:
                 while proceed():
                     try:
-                        r = OpenStack().run("volume show  -f json " +
-                                            volume_name)
-                        status = self.get_value(json.loads(r), 'status')
-                        if status == 'available':
-                            break
-                        else:
-                            log.info("volume " + volume_name +
-                                     " not available yet")
-                    except subprocess.CalledProcessError:
-                            log.info("volume " + volume_name +
-                                     " not information available yet")
-            # do not use OpenStack().run because its
-            # bugous for volume
-            with safe_while(sleep=20, increment=20, tries=10,
-                            action="add volume " + volume_name) as proceed:
-                while proceed():
-                    try:
-                        misc.sh("openstack server add volume " + name + " " + volume_name)
+                        self._create_volume(volume_name, volumes['size'])
+                        self._await_volume_available(volume_name)
                         break
-                    except subprocess.CalledProcessError:
-                        log.warning("openstack add volume failed unexpectedly; retrying")
+                    except MaxWhileTries:
+                        log.warning("openstack volume create+attach failed unexpectedly; retrying")
+            self._attach_volume(volume_name, name)
```

Comment:
The _attach_volume invocation should be in try loop, and on exceptions there should be cleanup steps for the volume in order to retry volume creation if volume exists but have some error status.